discovery-infra/test_infra/controllers/iptables.py

```diff
@@ -0,0 +1,44 @@
+import logging
+
+from test_infra.utils import run_command
+
+class IptableRule:
+    def __init__(self, chain, target, protocol, dest_port='', sources=None, extra_args=''):
+        self._chain = chain
+        self._target = target
+        self._protocol = protocol
+        self._dest_port = dest_port
+        self._sources = sources if sources else []
+        self._extra_agrs = extra_args
+
+    def _build_command_string(self, option):
+        sources_string = ','.join(self._sources)
+        rule = f'iptables --{option}'
+        rule += f' {self._chain}'
+        rule += f' -p {self._protocol}'
+        rule += f' -s {sources_string}' if self._sources else ''
+        rule += f' --dport {self._dest_port}' if self._dest_port else ''
+        rule += f' {self._extra_agrs}' if self._extra_agrs else ''
+        rule += f' -j {self._target}'
+
+        return rule
+    
+    def _does_rule_exist(self):
+        check_rule = self._build_command_string('check')
+        res = run_command(check_rule, shell=True, raise_errors=False)
```

Comment:
Assuming the rule doesn't exist, than you apply it. no?

Reply:
this function just checks if the rule exists

This is checked only on delete

You're right will check on insert as well